### PR TITLE
PUBDEV-5456 groupby rank

### PR DESCRIPTION
**Pull request by wendycwong**
For a H2OFrame, wants to achieve the following:
1. Groups are generated from some groupby columns;
2. Within each group of the groupby groups, sort the rows with some sort columns.
3. Generate a rank that is the sort order within the groupby group.  This rank is recorded in a new column attached to the H2OFrame.
4.  The final returned H2OFrame with the additional frame should be sorted according to the sort columns.

I have achieved this through 3 pass through the data using existing H2O functions as well:
1. Sort the H2OFrame use the sort columns and sort directions specified by the users.  Add new columns to store the rank data.
2. For each chunk of data, collect information: how many groupby groups (specified by the groupby columns) are there and for each groupby group, how many rows belong to that group.
3. For each chunk of data, assign rank to each row with information gathered about each chunk in step 2.

###########################
The above approach does not work well if the groupby columns are unique and they are as many groupby groups as the number of rows in the dataset.  In particular, I have created a big dataset and the memory blow up.  This test was suggested by Michal Kurka who worried about this memory blow up.  In coding, the trade off between execution speed and memory continues.  However, in this case, the memory criteria is what breaks first.

Instead, the following is done:
1. Sort the frame according to the groupby cols, sort cols in the direction specified by the user.  Groupby cols are always sorted ascending.
2. collect chunk information on the first and last groupby groups.
3. Assigned a rank to each row.
4. If the user wants, we can return the final frame sorted according to the sort cols and directions.  Default setting is to avoid this sort.  But the user can force it.

The worst case here is 4 pass over the dataset.  However, we save memory to only need to remember 2 pieces of information per chunk (first group by group and last groupby group).

I have added Java/Python/R unit tests to check and make sure this thing works.

APIs:
java: via rapids: rank_within_groupby groupby_cols_list sort_col_list sort_direction_list newColName sort_cols_order
Python: aH2OFrame.rankWithinGroupBy(groupCols_list, sortCols_list, sortDirs_list, newColName, sortColsSorted), the last three arguments are optional
R: h2o.rankWithinGroupBy(aH2OFrame, groupCols_list, sortCols_list, sortDirs_list, newColName, sortColsSorted), the last three arguments are optional

### CODE REVIEW

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstRankWithinGroupBy.java [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    }
+  }
+
+  public class SortnGrouby extends MRTask<SortnGrouby> {
+    final int[] _sortCols;
+    final int[] _groupbyCols;
+    final int[] _sortOrders;
+    final String _newColname;
+    Frame _groupedSortedOut;  // store final result
+    IcedHashMap<GChunk, String>[] _chunkStasG;  // store all groupby class per chunk
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstRankWithinGroupBy.java]: Rename GChunk to something else

**Comment by michalkurka** [comment]
Please add a test that will make sure the code doesn't crash when there is just a single group in the whole (large) dataset.

**Comment by wendycwong** [comment]
Need to do the following from Michal K feedback:
1. Make a test to test just one big group.
2. Run with huge dataset on HDFS, check the time to finish this and also the time to finish just sort with all groupby and sortcolumns.

**Comment by wendycwong** [comment]
Changed GChunk to GInfoPC.  Added big pyunit test with 6million rows of one groupby value only.  Done!

File: h2o-py/h2o/frame.py [hunk 2]
```diff
@@ -2577,6 +2577,42 @@ def pivot(self, index, column, value):
...
+        :param groupbyCols: The columns to group on (either a single column name/index, or a list of column names
+          or column indices
+        :param sortCols: The columns to sort on (either a single column name/index, or a list of column names or
+          column indices
+        :param ascending: Optional Boolean array to denote sorting direction for each sorting column.  True for
+          ascending, False for descending.  Default is ascending sort.  Sort direction for enums will be ignored.
+        :param newColName: Optional String to denote the new column names.  Default to New_Rank_column.
+
+        :return: a new Frame sorted by columns in sortCols with new rank column within the grouping specified
+          by the groupbyCols.
```
Comment (angela0xdata) [inline h2o-py/h2o/frame.py]: Would it be possible to include an example?

File: h2o-r/h2o-package/R/frame.R [hunk 3]
```diff
@@ -3782,6 +3782,67 @@ h2o.arrange <- function(x, ...) {
...
+
+  if(group.cols <= 0L || group.cols > ncol(data)) {
+   stop('Column ', group.cols, ' out of range for frame columns ', ncol(data), '.')
+  }
+
+  if (anyDuplicated(by)) stop("Some duplicate column names have been provided")
+
+  return(group.cols)
+}
+
```
Comment (angela0xdata) [inline h2o-r/h2o-package/R/frame.R]: The first comment line should be a short summary, like "Ranks rows within a GroupBy object" (or whatever makes sense). Then this becomes the function description in the docs. Right now, this shows up as a very large paragraph for the function summary, and the "Description" is only the "To sort column c1..." paragraph. (Which could be a good paragraph for the "Details" section.)

File: h2o-r/h2o-package/R/frame.R [hunk 4]
```diff
@@ -3782,6 +3782,67 @@ h2o.arrange <- function(x, ...) {
...
+#' First, sorts an H2OFrame by columns in sortby and sort directions specified.
+#' Next, groupby groups are formed by the columns specified in by.
```
Comment (angela0xdata) [inline h2o-r/h2o-package/R/frame.R]: "Next, groupby groups are formed by the columns specified in by." 
Should that be:
"Next, groupby groups are formed by the columns specified in groupbyCols" instead? (I don't see "by" as a parameter.)

File: h2o-py/h2o/frame.py (hunk 2 above)
Comment (wendycwong) [reply h2o-py/h2o/frame.py]: This is a good idea.  Yes, Nidhi has an example.  Let me asked if she will let me include it.

File: h2o-r/h2o-package/R/frame.R (hunk 4 above)
Comment (wendycwong) [reply h2o-r/h2o-package/R/frame.R]: Angela:  Let me redo this and make it better for you.   Stay tuned....

**Comment by wendycwong** [comment]
For performance test, I have run with a big dataset of 4 columns and 630000000 rows.  The tests were run on a hadoop cluster of 4 with 40G.  Here are the test results in seconds:

sort over all 4 columns, no ranking
53.037
54.427
58.711 
54.941 
52.518 

rank-within-group-by without final resort
60.205 
55.319
56.785
54.18
53.258
54.426

rank-within-group-by with final resort
107.441 
103.66 
106.37 
110.337
103.141 

Observations:
1. Ranking is very light weight.  It does not add much to the execution time.
2. As you can see, skipping the final sort can shave about 15 seconds off the execution time.

File: h2o-py/h2o/frame.py [hunk 5]
```diff
@@ -2577,6 +2577,128 @@ def pivot(self, index, column, value):
...
+            returned as is with no extra sort.  This may provide a small speedup if desired.
+
+        :param group_by_cols: The columns to group on (either a single column name/index, or a list of column names
+          or column indices
+        :param sort_cols: The columns to sort on (either a single column name/index, or a list of column names or
+          column indices
+        :param ascending: Optional Boolean array to denote sorting direction for each sorting column.  True for
+          ascending, False for descending.  Default is ascending sort.  Sort direction for enums will be ignored.
+        :param new_col_name: Optional String to denote the new column names.  Default to New_Rank_column.
+        :param sort_cols_sorted: Optional Boolean to denote if the returned frame should be sorted accoring to sort_cols
```
Comment (ledell) [inline h2o-py/h2o/frame.py]: typo: "accoring" should be "according"

Comment (wendycwong) [reply h2o-py/h2o/frame.py]: Fixed the typo and the assignment in python from <- to =.  Thanks, Erin.

File: h2o-py/h2o/frame.py [hunk 6]
```diff
@@ -2577,6 +2577,128 @@ def pivot(self, index, column, value):
...
+         16               3 -22905.288    0                    3 -61284.6900     0
+         17               2 -13465.496    1                    2  12094.4851     1
+         17               2 -13465.496    1                    3 -11772.1338     1
+         17               2 -13465.496    1                    3   -415.1114     0
+         17               2  -3329.619    1                    2  12094.4851     1
+         17               2  -3329.619    1                    3 -11772.1338     1
+         17               2  -3329.619    1                    3   -415.1114     0
+
+         If the following commands are issued:
+         rankedF1 = h2o.rank_within_group_by(train, ["Group_by_column"], ["Column_to_arrange_by"], [TRUE])
```
Comment (ledell) [inline h2o-py/h2o/frame.py]: same issue as before, should be `c("Group_by_column")`

File: h2o-r/h2o-package/R/frame.R [hunk 7]
```diff
@@ -3782,6 +3782,149 @@ h2o.arrange <- function(x, ...) {
   .newExpr("sort", x, by-1L, as.numeric(ascend))
 }
 
+generateColInd <-function(data, by) {
```
Comment (ledell) [inline h2o-r/h2o-package/R/frame.R]: is this an internal function?  It's not documented and exported (so it won't show up in R documentation).  Either way, probably change the name to `generate_col_index`

File: h2o-r/h2o-package/R/frame.R (hunk 3 above)
Comment (ledell) [reply h2o-r/h2o-package/R/frame.R]: @wendycwong See my comment above, same as @angela0xdata.

**Pull request merged**